src/platform/packages/shared/kbn-esql-ast/src/definitions/utils/validation/function.ts

```diff
@@ -921,4 +951,189 @@ export function isSupportedFunction(
+ * // Returns ["ST_EXTENT_AGG with parameter 'field' of type 'cartesian_shape' requires a PLATINUM license."]
+ */
+function validateSignatureLicense(
+  fn: ESQLFunction,
+  matchingSignatures: FunctionDefinition['signatures'],
+  hasMinimumLicenseRequired: (minimumLicenseRequired: ESQLLicenseType) => boolean,
+  context: ICommandContext,
+  parentCommand: string
+): ESQLMessage[] {
+  if (matchingSignatures.length === 0) {
+    return [];
+  }
+
+  // Get signatures that match argument types
+  const typeMatchingSignatures = getValidationTypeMatchingSignatures(
+    fn,
+    matchingSignatures,
+    context,
+    parentCommand
+  );
+
+  // This ensures we only validate licenses for signatures that could actually be used
+  // Note: matchingSignatures is a fallback for cases like ST_EXTENT_AGG(wrongField).
+  // example: ST_EXTENT_AGG(0) → typeMatchingSignatures = [], but hasUnlicensedSignature must be true to bypass the license check and to to the next step.
+  const relevantSignatures =
+    typeMatchingSignatures.length > 0 ? typeMatchingSignatures : matchingSignatures;
+
+  // If the user can use at least one signature without a license, the function is allowed
+  const hasUnlicensedSignature = relevantSignatures.some((sig) => !sig.license);
+  if (hasUnlicensedSignature) {
+    return [];
+  }
+
+  // If the user has the required license for at least one signature, allow the function
+  const hasValidLicense = relevantSignatures.some(
+    (signature) =>
+      signature.license &&
+      hasMinimumLicenseRequired(signature.license.toLocaleLowerCase() as ESQLLicenseType)
+  );
+
+  if (hasValidLicense) {
+    return []; // License validation passes
+  }
+
+  // Generate a specific error message indicating which signature requires the license
+  const firstLicensedSignature = relevantSignatures.find((sig) => sig.license);
+  const requiredLicense = firstLicensedSignature?.license?.toUpperCase() || 'UNKNOWN';
+
+  const signatureDescription = generateSignatureDescription(firstLicensedSignature);
+
+  return [
+    getMessageFromId({
+      messageId: 'licenseRequiredForSignature',
+      values: {
+        name: fn.name.toUpperCase(),
+        signatureDescription,
+        requiredLicense,
+      },
+      locations: fn.location,
+    }),
+  ];
+}
+
+/**
+ * Helper function to filter signatures that match argument types for validation.
+ * Similar to getValidFunctionSignaturesForPreviousArgs but for complete validation
+ * with full context and array handling.
+ */
+function getValidationTypeMatchingSignatures(
+  fn: ESQLFunction,
+  signatures: FunctionDefinition['signatures'],
+  context: ICommandContext,
+  parentCommand: string
+): FunctionDefinition['signatures'] {
+  return signatures.filter((signature) =>
+    signature.params.every((param, index) => {
+      const arg = fn.args[index] as ESQLSingleAstItem | ESQLSingleAstItem[];
+
+      // Handle missing arguments
+      if (!arg) {
+        return param.optional;
```

Comment:
In theory I haven't seen this case and typescript doesn't warn me, but I put this check in just in case